collections/_tutorials/2022-06-20-tutorial_contrastive_learning.md

```diff
@@ -0,0 +1,117 @@
+
+* A neural network base encoder $$f(\cdot)$$ that extracts representation vectors from augmented data samples. This step can be realized by any standard network architecture, such as ResNet. This step is modeled as follows:
+
+$$h_i = f(\tilde{x}_i)=\texttt{ResNet}(\tilde{x}_i)$$
+
+&nbsp; &nbsp; &nbsp; &nbsp; where $$h_i \in \mathbb{R}^{d}$$ is the output vector after average pooling layer. 
+
+* A simple neural network projection head $$g(\cdot)$$ that maps representations to the space where the contrastive loss is applied. A simple MLP with one hidden layer is  used for this step which outputs a vector $$z_i \in \mathbb{R}^{d'}$$ with $$d'=128$$. This allows to project the representation to a 128-dimensional latent space.
+
+* A contrastive loss function defined for a contrastive prediction task. Given a set of $$\{\tilde{x}_k\}$$ including a positive pair of samples $$\tilde{x}_i$$ and $$\tilde{x}_j$$, the contrastive prediction task aims to identify $$\tilde{x}_j$$ in $$\{\tilde{x}_k\}_{k \neq i}$$ for a given $$\tilde{x}_i$$.
+
+&nbsp;
+
+### Contrastive loss
+
+* A minibatch of $$N$$ samples is randomly selected.
+
+* Each sample is used to create a pair of augmented examples, resulting in $$2N$$ data points.
+
+* For each minibatch, one pair is considered as positive and the others $$2(N-1)$$ as negative examples. 
+
+* The cosine similarity between two samples $$u$$ and $$v$$ is computed from the conventional dot product $$sim(u,v)=u^{T}v/\left(\|u\|\|v\|\right)$$ . The corresponding value varies between 0 (when u and v are orthogonal) and 1 (when u and v are aligned).
+
+* The following loss function for a positive pair of samples $$(i,j)$$ is defined as:
+
+$$l(i,j)=-\log{\left(\frac{ \exp\left(sim(z_i,z_j)/\tau\right) }{ \sum_{k=1}^{2N}{\mathbb{1}_{[k \neq i]}\exp\left(sim(z_i,z_k)/\tau\right) } }\right)}$$
+
+&nbsp; &nbsp; &nbsp; &nbsp; where $$\mathbb{1}_{[k \neq i]} \in [0,1]$$ is an indicator function evaluating to 1 iff $$k \neq i$$ and $$\tau$$ is a parameter.
+
+>> Since both the numerator and denominator involve exponential terms, the values inside the $$-\log(\cdot)$$ function varie between 0 and 1. As a reminder, we display above the corresponding curve. From this figure, on can see that the loss function will tend to its minimum when the numerator and the denominator will be close, i.e. when the set $$\{\exp\left(sim(z_i,z_k)/\tau\right)\}_{[k \neq (i,j)]}$$ will be as low as possible and when $$\exp\left(sim(z_i,z_j)/\tau\right)$$ will be high, making the two points $$z_i$$ and $$z_j$$ to be as close as possible and in the meanwhile the other points to be as orthogonal as possible to these two points. The minimization of the contrastive loss thus allows to structure the latent space according to visual representation.
```

Comment:
Typo + rephrase the long sentence.
The figure is below?
```suggestion
>> Since both the numerator and denominator involve exponential terms, the values inside the $$-\log(\cdot)$$ function vary between 0 and 1. As a reminder, the corresponding curve is plotted right below. On this figure, we can see that the loss function tends to its minimum when the numerator and the denominator are close. This means the set $$\{\exp\left(sim(z_i,z_k)/\tau\right)\}_{[k \neq (i,j)]}$$ is as low as possible and $$\exp\left(sim(z_i,z_j)/\tau\right)$$ is high, making the two points $$z_i$$ and $$z_j$$ to be as close as possible while the other points are as orthogonal as possible to these two points. Thus, the minimization of the contrastive loss allows to structure the latent space according to the visual representation.
```